File: 22-generate-parentheses/22-generate-parentheses.cpp
class Solution {
public:
     vector<string> v;
    vector<string> generateParenthesis(int n) {
        string s;
        generate(s,n,n);
        return v;
    }
    void generate(string &s,int op,int cl){
       if(op ==0 && cl==0){
           v.push_back(s);
           return ;
       }
        if(op>0){
            s.push_back('(');
            generate(s,op-1,cl);
            s.pop_back();
        }
        if(cl>0){
            if(op<cl){
                s.push_back(')');
                generate(s,op,cl-1);
                s.pop_back();
            }
        }
    }
};

File: 22-generate-parentheses/22-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> gen_paren = new ArrayList<>();
        backtrack(gen_paren,"",0,0,n);
        return gen_paren;
    }
    public void backtrack( List<String> gen_paren,String s,int open,int close,int n){
        if(s.length() == n*2){
            gen_paren.add(s);
            return ;
        }
        if(open<n)
            backtrack(gen_paren,s+"(",open+1,close,n);
        if(close<open)
            backtrack(gen_paren,s+")",open,close+1,n);
    }
}